### PR TITLE
posix: device_io: implement fdopen(), fileno(), pread(), pselect(), pwrite()

### LINKED ISSUE #24966: sockets/posix: poll()/select() implementation should be factored out from sockets subsys into POSIX subsys
POSIX poll() was initially implemented in the scope of BSD Sockets API, and currently located in subsys/net/lib/sockets/ . It is however a generic POSIX facility, and was generalized to apply to arbitrary file descriptors (and objects represented by them). As we grow implementations of such objects (eventfd, unix domain sockets), dependency on networking sockets becomes a problem, as requires enabling unrelated config options and leads to pulling more code into user applications.

To address this, poll() (and by extension, select()) implementation should be migrated to lib/posix/.

### LINKED ISSUE #66932: posix: device_io: implement `fdopen()`
This is part of the See [POSIX Roadmap for LTSv3](https://github.com/zephyrproject-rtos/zephyr/issues/51211) (RFC #51211).

`fdopen()` is required as part of `POSIX_DEVICE_IO` [Subprofiling Option Group](https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_subprofiles.html).

For more information, please refer to https://pubs.opengroup.org/onlinepubs/9699919799/functions/fdopen.html

### LINKED ISSUE #66938: posix: device_io: implement `fileno()`
This is part of the See [POSIX Roadmap for LTSv3](https://github.com/zephyrproject-rtos/zephyr/issues/51211) (RFC #51211).

`fileno()` is required as part of `POSIX_DEVICE_IO` [Subprofiling Option Group](https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_subprofiles.html).

For more information, please refer to https://pubs.opengroup.org/onlinepubs/9699919799/functions/fileno.html

### LINKED ISSUE #66946: posix: device_io: implement `pread()`
This is part of the See [POSIX Roadmap for LTSv3](https://github.com/zephyrproject-rtos/zephyr/issues/51211) (RFC #51211).

`pread()` is required as part of `POSIX_DEVICE_IO` [Subprofiling Option Group](https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_subprofiles.html).

For more information, please refer to https://pubs.opengroup.org/onlinepubs/9699919799/functions/pread.html

### LINKED ISSUE #66947: posix: device_io: implement `pselect()`
This is part of the See [POSIX Roadmap for LTSv3](https://github.com/zephyrproject-rtos/zephyr/issues/51211) (RFC #51211).

`pselect()` is required as part of `POSIX_DEVICE_IO` [Subprofiling Option Group](https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_subprofiles.html).

For more information, please refer to https://pubs.opengroup.org/onlinepubs/9699919799/functions/pselect.html

### LINKED ISSUE #66948: posix: device_io: implement `pwrite()`
This is part of the See [POSIX Roadmap for LTSv3](https://github.com/zephyrproject-rtos/zephyr/issues/51211) (RFC #51211).

`pwrite()` is required as part of `POSIX_DEVICE_IO` [Subprofiling Option Group](https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_subprofiles.html).

For more information, please refer to https://pubs.opengroup.org/onlinepubs/9699919799/functions/pwrite.html

### PR DESCRIPTION
**Pull request by cfriedt**
This is a resubmission of #73978

Implement the remaining functions in the `POSIX_DEVICE_IO` Option Group.

This Option Group is requried for the PSE51, PSE52, PSE53, and PSE54 subprofiles and likely others.

[Doc Preview](https://builds.zephyrproject.io/zephyr/pr/76352/docs/services/portability/posix/index.html)

Fixes #24966
Fixes #66932 
Fixes #66938 
Fixes #66946 
Fixes #66947 
Fixes #66948

### CODE REVIEW

**Comment by cfriedt** [comment]
I think this is basically ready for review. It includes the less intrusive workaround of defining off_t if it isn't defined in a very small corner case.

The posix - zephyr - posix dependency cycle has been separated out to it's own issue so that it doesn't block this feature.

**Comment by cfriedt** [comment]
* formatted to silence a clang-format warning

**Comment by cfriedt** [comment]
Kicking CI again, as most jobs were cancelled

**Review by ycsin — DISMISSED** [review]
LGTM

File: lib/posix/options/Kconfig.fs [hunk 1]
```diff
@@ -15,7 +15,7 @@ if POSIX_FILE_SYSTEM
 config POSIX_FILE_SYSTEM_ALIAS_FSTAT
 	bool
 	help
-	  Select 'y' here and Zephyr will provide an alias for fstat() as _fstat().
+	  When selected via Kconfig, Zephyr will provide an alias for fstat() as _fstat().
```
Comment (fgrandel) [inline lib/posix/options/Kconfig.fs:18]:
```suggestion
	  When selected, Zephyr will provide an alias for fstat() as _fstat().
```

File: lib/os/zvfs/Kconfig [hunk 2]
```diff
@@ -33,4 +33,27 @@ config ZVFS_EVENTFD_MAX
...
 
+config ZVFS_POLL
+	bool "ZVFS poll"
+	select POLL
+	help
+	  Enable support for zvfs_poll().
+
+if ZVFS_POLL
+
+config ZVFS_POLL_MAX
```
Comment (rlubos) [inline lib/os/zvfs/Kconfig:44]: Don't we need to deprecate `NET_SOCKETS_POLL_MAX` along with this patch? The config still exists, is being set by several samples/tests, but is no longer effective. I recall we've discussed this during the previous attempt: https://github.com/rlubos/zephyr/commit/6c77edc20519439d5305f678582c13c017172e40

**Review by rlubos — COMMENTED** [review]
Generally looks fine, but we need to sort out the symbol deprecation.

File: lib/posix/options/Kconfig.fs (hunk 1 above)
Comment (cfriedt) [reply lib/posix/options/Kconfig.fs:18]: This one is non-user-selectable, so it needs to be done via Kconfig (typically by some libc-specific Kconfig).

File: lib/os/zvfs/Kconfig (hunk 2 above)
Comment (cfriedt) [reply lib/os/zvfs/Kconfig:44]: Would it be OK if I cherry-pick your commit?

Comment (cfriedt) [reply lib/os/zvfs/Kconfig:44]: K - cherry-picked - I'm hoping CI doesn't complain 😅

**Comment by cfriedt** [comment]
* deprecated `CONFIG_NET_SOCKETS_POLL_MAX` in favour of `CONFIG_ZVFS_POLL_MAX`
* rebased on `main`

File: lib/os/zvfs/Kconfig (hunk 2 above)
Comment (rlubos) [reply lib/os/zvfs/Kconfig:44]: sure, no prob, not sure though if it's not outdated. I guess CI will tell us

**Comment by cfriedt** [comment]
* fixed duplicate entry of `CONFIG_ZVFS_POLL_MAX` by adding `CONFIG_ZVFS_OPEN_MAX`

**Review by rlubos — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**